cdc/puller/sorter/heap_sorter.go

```diff
@@ -284,6 +319,35 @@ func (h *heapSorter) init(ctx context.Context, onError func(err error)) {
 	h.internalState = state
 }
 
+// asyncCanceller is a shared object used to cancel async IO operations.
+// We do not use `context.Context` because (1) selecting on `ctx.Done()` is expensive
+// and costly especially if the context is shared by many goroutines (2) due to the complexity
+// of managing contexts through the workerpools, using a special shared object seems more reasonable
+// and readable.
+type asyncCanceller struct {
+	exitRWLock sync.RWMutex // held when an asynchronous flush is taking place
+	hasExited  bool         // protected by exitRWLock
+}
+
+func (c *asyncCanceller) EnterAsyncOp() {
+	c.exitRWLock.RLock()
+}
+
+func (c *asyncCanceller) FinishAsyncOp() {
+	c.exitRWLock.RUnlock()
+}
+
+func (c *asyncCanceller) IsCanceled() bool {
+	return c.hasExited
+}
```

Comment:
Can we remove `EnterAsyncOp` and `FinishAsyncOp` by rewriting `IsCanceled` to the following?

```suggestion
func (c *asyncCanceller) IsCanceled() bool {
	c.exitRWLock.Lock()
	defer c.exitRWLock.Unlock()

	return c.hasExited
}
```

Reply:
The whole IO operation needs to be protected so that we don't try to release the files when they are being written to. So it's advisable to wrap the entire execution in the read lock.